crates/runtime/src/instance/allocator/pooling/uffd.rs

```diff
@@ -0,0 +1,833 @@
+                            )
+                        })?;
+                }
+                _ => {
+                    log::trace!("zeroing page at {:p} with length {}", dst, page_size);
+
+                    // No data, zero the page without waking
+                    uffd.zeropage(dst as _, *page_size, false).map_err(|e| {
+                        format!(
+                            "failed to zero page at {:p} with length {}: {}",
+                            dst, page_size, e
+                        )
+                    })?;
+                }
+            }
+        }
+
+        // Finally wake the entire wasm page
+        uffd.wake(page_addr as _, WASM_PAGE_SIZE).map_err(|e| {
+            format!(
+                "failed to wake page at {:p} with length {}: {}",
+                page_addr, WASM_PAGE_SIZE, e
+            )
+        })
+    } else {
+        log::trace!(
+            "initialization data is not paged; zeroing Wasm page at {:p}",
+            page_addr
+        );
+
+        uffd.zeropage(page_addr as _, WASM_PAGE_SIZE, true)
+            .map_err(|e| {
+                format!(
+                    "failed to zero page at {:p} with length {}: {}",
+                    page_addr, WASM_PAGE_SIZE, e
+                )
+            })?;
+
+        Ok(())
+    }
+}
+
+fn handler_thread(
+    uffd: Uffd,
+    locator: AddressLocator,
+    mut registrations: usize,
+    faulted_stack_guard_pages: Arc<[AtomicBool]>,
+) -> Result<(), String> {
+    loop {
+        match uffd.read_event().expect("failed to read event") {
+            Some(Event::Unmap { start, end }) => {
+                log::trace!("memory region unmapped: {:p}-{:p}", start, end);
+
+                let (start, end) = (start as usize, end as usize);
+
+                if (start == locator.memories_start && end == locator.memories_end)
+                    || (start == locator.tables_start && end == locator.tables_end)
+                    || (start == locator.stacks_start && end == locator.stacks_end)
+                {
+                    registrations -= 1;
+                    if registrations == 0 {
+                        break;
+                    }
+                } else {
+                    panic!("unexpected memory region unmapped");
+                }
+            }
+            Some(Event::Pagefault {
+                addr: access_addr, ..
+            }) => {
+                unsafe {
+                    match locator.get_location(access_addr as usize) {
+                        Some(AddressLocation::TablePage { page_addr, len }) => {
+                            log::trace!(
+                                "handling fault in table at address {:p} on page {:p}",
+                                access_addr,
+                                page_addr,
+                            );
+
+                            // Tables are always initialized upon instantiation, so zero the page
```

Comment:
I think I now see what you meant.

`Table::new_dynamic` (previously `Table::new`) is explicitly using `ptr::null_mut` and `None` (which is an `Option<NotNull<T>>`) for "uninitialized" elements, and this will assume those are equivalent to a zeroed page when uffd is in play.

I'll add comments in `Table` that says there's an assumption being made here that uninitialized elements are assumed to be represented with all zeroes.

Reply:
`Table` now stores by raw pointers for both "static" and "dynamic" storage, so these assumptions are now being enforced via `TableElement`.